### PR TITLE
Added that subqueries related to another database work

### PR DESCRIPTION
**Pull request by maguilar92**
#23003 

Added cross database compatibility in subqueries including database name in subquery when query connection and subquery connection are different.

# COMPATIBILITY

- MySQL (Compatible)
- PostgreSQL (Compatible)
- SQLite (Not compatible) **will work as it works now**
- SQL Server (Compatible)

This will not break any existing features because user no need to modify existing code. 
Only the subqueries with table in another database incorporate the database name to the left of the table name.

# BENEFITS
Users can use the methods has, whereHas, doesntHave and whereDoesntHave in Eloquent when the relation is in another database.

For example:

`Provider::whereHas('orders')`

### NOW GENERATES

````select * from `hvp_providers` where exists (select * from `hvp_products` where `hvp_providers`.`id` = `hvp_products`.`provider_id` and `hvp_products`.`deleted_at` is null) and `hvp_providers`.`deleted_at` is null````

**Sql error:** 
SQLSTATE[42S02]: Base table or view not found: 1146 Table 'global.hvp_products' doesn't exist

### WITH THIS PULL REQUEST GENERATES

````select * from `hvp_providers` where exists (select * from `shop`.`hvp_products` where `hvp_providers`.`id` = `hvp_products`.`provider_id` and `hvp_products`.`deleted_at` is null) and `hvp_providers`.`deleted_at` is null````

### CODE REVIEW

File: src/Illuminate/Database/Eloquent/Concerns/QueriesRelationships.php [hunk 1]
```diff
@@ -47,11 +47,32 @@ public function has($relation, $operator = '>=', $count = 1, $boolean = 'and', C
...
         return $this->addHasWhere(
             $hasQuery, $relation, $operator, $count, $boolean
         );
     }
 
+    protected function hasCanCrossDatabase($hasQuery)
+    {
+        foreach (['MySqlConnection', 'PostgresConnection', 'SqlServerConnection'] as $connector) {
+            if (is_a($this->getConnection(), '\Illuminate\Database\\'.$connector) && is_a($hasQuery->getConnection(), '\Illuminate\Database\\'.$connector)) {
+                return true;
```
Comment (sisve) [inline src/Illuminate/Database/Eloquent/Concerns/QueriesRelationships.php]: This looks problematic; what if I have a custom database connection? Would it be possible to check for an interface instead? `return $this->getConnection() instanceof CanCrossDatabaseShazaamInterface;`

File: src/Illuminate/Database/Query/Grammars/Grammar.php [hunk 2]
```diff
@@ -140,6 +140,13 @@ protected function compileColumns(Builder $query, $columns)
      */
     protected function compileFrom(Builder $query, $table)
     {
+        // if table contains | means that we must specify the database
+        if (strpos($table, '|') !== false) {
```
Comment (sisve) [inline src/Illuminate/Database/Query/Grammars/Grammar.php]: What if someone has a legitimate table with a pipe character in the name?

File: src/Illuminate/Database/Eloquent/Concerns/QueriesRelationships.php [hunk 3]
```diff
@@ -47,11 +47,32 @@ public function has($relation, $operator = '>=', $count = 1, $boolean = 'and', C
             $hasQuery->callScope($callback);
         }
 
+        // When query connection and subquery connection are not equals concat connection to query from
```
Comment (jmarcher) [inline src/Illuminate/Database/Eloquent/Concerns/QueriesRelationships.php]: Comments should follow the Laravel style

File: src/Illuminate/Database/Query/Grammars/Grammar.php [hunk 4]
```diff
@@ -140,6 +140,13 @@ protected function compileColumns(Builder $query, $columns)
      */
     protected function compileFrom(Builder $query, $table)
     {
+        // if table contains | means that we must specify the database
```
Comment (jmarcher) [inline src/Illuminate/Database/Query/Grammars/Grammar.php]: Comments should follow the Laravel style

File: src/Illuminate/Database/Eloquent/Concerns/QueriesRelationships.php (hunk 1 above)
Comment (jmarcher) [reply src/Illuminate/Database/Eloquent/Concerns/QueriesRelationships.php]: I like sisve's idea more

**Comment by taylorotwell** [comment]
Too complicated won't be merged.

**Comment by maguilar92** [comment]
@taylorotwell What do I have to do to implement the change? What solution is there for this type of case?

**Pull request closed without merging**